### DISCUSSION TITLE
Response pass through Psr/StreamInterface or Http client response

### DISCUSSION THREAD

**Discussion by SEUH** [laravel/framework#49991] [Ideas]
For Http or Guzzle responses that exceed the memory limit (or if calling `sendContent()` exceeds it due to copying the response `$content`) you want to pass through the responses directly to the user but atm. you have to either save the response to a tmpfile or use the callback() function from Response::stream. In every one of my projects i have some internal files/apis that i need to directly pipe through and their responses exceed "normal memory limits".

What i currently do with a macro:
```php
return Response::passThrough(Http::get('http://some-api.internal'))
```

I would like to add two more functions to `src/Illuminate/Routing/ResponseFactory.php` but i wouldn't add a `instanceof` check inside `stream()` or `streamDownload()`. I'm also not sure how to name them. Feel free to suggest and if this is useful i can create a PR.

I intentionally chose to loop over `$stream` instead of using `fpassthru` because that function has it's own problems and i haven't tested every `StreamInterface` implementation.

```php
// src/Illuminate/Routing/ResponseFactory.php
public function passThrough(\Illuminate\Http\Client\Response|StreamInterface $stream, $status = 200, array $headers = [])
{
    return $this->stream(
        $this->callbackForStream($this->resolveStream($stream)),
        $status,
        $headers,
    );
}

public function passThroughDownload(\Illuminate\Http\Client\Response|StreamInterface $stream, $name = null, array $headers = [], $disposition = 'attachment')
{
    return $this->streamDownload(
        $this->callbackForStream($this->resolveStream($stream)),
        $name,
        $headers,
        $disposition,
    );
}

private function resolveStream(\Illuminate\Http\Client\Response|StreamInterface $stream)
{
    if ($stream instanceof \Illuminate\Http\Client\Response) {
        $stream = $stream->toPsrResponse()->getBody();
    } else if ($stream instanceof \Psr\Http\Message\MessageInterface) {
        $stream = $stream->getBody();
    }

    return $stream;
}

private function callbackForStream(\Illuminate\Http\Client\Response|StreamInterface $stream)
{
    return function () use ($stream) {
        while (!$stream->eof()) {
            echo $stream->read(1024 * 1024);
        }
        $stream->close();
    };
}
```

**Comment by ghost** [comment]
Hi. What is the issue with the callback that is now implemented?
return \response()->stream(... ?

**Reply by SEUH** [reply]
There is no issue. This is a "feature request". I have this code as a macro in every one of my projects. These two functions are helpfull to simply do this

```php
return Response::passThrough(Http::get('http://some-internal-api.com'))
```
I know you can do this
```php
return Response::make(Http::get('http://some-internal-api.com')->body());
```

but the problem is that it duplicated the content size of the get body. So if you want to forward response bodys bigger than half of your memory_limit you need to write a chunked pass through function similar to mine.

**Reply by ghost** [reply]
I understand. I see a pattern here in the discussions. Useful ideas requested to be implemented in laravel (increasing the overhead) that have only 1 problem. They are not used by the majority of people using laravel. I also have code that I overwrite using macros etc. but that is because my specific case needs it. If you have duplicate code in all your projects, why don't you create a package or a lib that you include in all of them? You can even make public that package if you want. At the end of the day, if your idea/package gets traction and is used by the majority of artisans, then they will include it into laravel.

**Comment by SMFloris** [comment]
@marius-mcp having a nice helper to passthrough a StreamInterface would be great since it is in PSR as well. We are using simillar macros accross our codebase as well, so it would be amazing if laravel could have this out of the box.
